iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/introspection/IamIntrospectionEndpoint.java

```diff
@@ -0,0 +1,266 @@
+    }
+  }
+
+  private IntrospectionResponse introspectRefreshToken(String token)
+      throws ParseException, InvalidTokenException {
+
+    OAuth2RefreshTokenEntity rt = tokenService.getRefreshToken(token);
+    if (rt.isExpired()) {
+      return IntrospectionResponse.inactive();
+    }
+    IntrospectionResponse.Builder builder = new IntrospectionResponse.Builder(true);
+    // base response: copy all claims
+    rt.getJwt().getJWTClaimsSet().getClaims().forEach(builder::addField);
+    // token type
+    builder.addField(OAuth2TokenIntrospectionClaimNames.TOKEN_TYPE, TokenTypeHint.REFRESH_TOKEN);
+    builder.addField(OAuth2TokenIntrospectionClaimNames.SUB,
+        rt.getJwt().getJWTClaimsSet().getSubject());
+    builder.addField(OAuth2TokenIntrospectionClaimNames.CLIENT_ID, rt.getClient().getClientId());
+    includeIfNotNull(builder, "client_name", rt.getClient().getClientName());
+    includeIfNotNull(builder, "client_last_used", rt.getClient().getClientLastUsed());
+    builder.addField(OAuth2TokenIntrospectionClaimNames.EXP, rt.getExpiration());
+    return builder.build();
+  }
+
+  private IntrospectionResponse introspectAccessToken(String token)
+      throws InvalidTokenException, ParseException {
+
+    OAuth2AccessTokenEntity at = tokenService.readAccessToken(token);
+    if (at.isExpired() || isRevoked(at)) {
+      return IntrospectionResponse.inactive();
+    }
+    IntrospectionResponse.Builder builder = new IntrospectionResponse.Builder(true);
+    // base response: copy all token claims
+    at.getJwt().getJWTClaimsSet().getClaims().forEach(builder::addField);
+    // token type
+    builder.addField(OAuth2TokenIntrospectionClaimNames.TOKEN_TYPE, TokenTypeHint.ACCESS_TOKEN);
+    // override/populate more claims
+    String subject = at.getJwt().getJWTClaimsSet().getSubject();
+    String clientId = at.getClient().getClientId();
+    builder.addField(OAuth2TokenIntrospectionClaimNames.SUB, subject);
+    builder.addField(OAuth2TokenIntrospectionClaimNames.CLIENT_ID, clientId);
+    includeIfNotNull(builder, "client_name", at.getClient().getClientName());
+    includeIfNotNull(builder, "client_last_used", at.getClient().getClientLastUsed());
+    builder.addField(OAuth2TokenIntrospectionClaimNames.EXP, at.getExpiration());
+    builder.addField(OAuth2TokenIntrospectionClaimNames.IAT,
+        at.getJwt().getJWTClaimsSet().getIssueTime());
+    builder.addField(OAuth2TokenIntrospectionClaimNames.ISS,
+        at.getJwt().getJWTClaimsSet().getIssuer());
+    builder.addField(OAuth2TokenIntrospectionClaimNames.JTI,
+        at.getJwt().getJWTClaimsSet().getJWTID());
+    includeIfNotNull(builder, OAuth2TokenIntrospectionClaimNames.NBF,
+        at.getJwt().getJWTClaimsSet().getNotBeforeTime());
+    includeIfNotEmpty(builder, OAuth2TokenIntrospectionClaimNames.AUD,
+        at.getJwt().getJWTClaimsSet().getAudience());
+    includeIfNotEmpty(builder, OAuth2TokenIntrospectionClaimNames.SCOPE, at.getScope());
```

Comment:
The token_scope table contains token scopes even if they do not appear in the token itself; maybe you could take the information from there?

Reply:
> The token_scope table contains token scopes even if they do not appear in the token itself;
That's true. But it's exactly the purpose of using at.getScope() and not using the claims of the token. This method reads the content of token_scope.